pkg/metrics/providers/datadog.go

```diff
@@ -0,0 +1,160 @@
+package providers
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+	"time"
+
+	flaggerv1 "github.com/weaveworks/flagger/pkg/apis/flagger/v1beta1"
+)
+
+// https://docs.datadoghq.com/api/
+const (
+	datadogDefaultHost = "https://api.datadoghq.com"
+
+	datadogMetricsQueryPath     = "/api/v1/query"
+	datadogAPIKeyValidationPath = "/api/v1/validate"
+
+	datadogAPIKeySecretKey = "datadog_api_key"
+	datadogAPIKeyHeaderKey = "DD-API-KEY"
+
+	datadogApplicationKeySecretKey = "datadog_application_key"
+	datadogApplicationKeyHeaderKey = "DD-APPLICATION-KEY"
+
+	fromDeltaMultiplierOnProgressiveDeadline = 10
+)
+
+// DatadogProvider executes datadog queries
+type DatadogProvider struct {
+	metricsQueryEndpoint     string
+	apiKeyValidationEndpoint string
+
+	timeout        time.Duration
+	apiKey         string
+	applicationKey string
+	fromDelta      int64
+}
+
+type datadogResponse struct {
+	Series []struct {
+		Pointlist [][]float64 `json:"pointlist"`
+	}
+}
+
+// NewDatadogProvider takes a canary spec, a provider spec and the credentials map, and
+// returns a Datadog client ready to execute queries against the API
+func NewDatadogProvider(cd *flaggerv1.Canary,
```

Comment:
Instead of the canary param, the provider should receive a param called `windowSize` the delta should be calculated from that. The window size is set with `metric.interval`.